### DISCUSSION TITLE
signing a PDF without a form inside

### DISCUSSION THREAD

**Discussion by ljoli** [MatthiasValvekens/pyHanko#226] [Q&A]
**Describe the bug**
Hello,  I take a PDF from the net : 
https://www.utc.fr/~jlaforet/Suppl/python-cheatsheets.pdf

And I try to make a signature on this PDF but I have the following  error  : 

```
pyhanko sign addfields --field  2/10,10,10,10/NewS python-cheatsheetsH.pdf  output-21.pdf
2023-02-21 14:48:47,454 - pyhanko.cli - ERROR - Generic processing error.
Error: Generic processing error.
```

**To Reproduce**

1. Take the PDF on the link above
2. Run the cmd : pyhanko sign addfields --field  2/10,10,10,10/NewS python-cheatsheetsH.pdf  output-21.pdf

**Expected behavior**
I want add empty field for add a signature on PDF.

**Environment (please complete the following information):**
 - OS: openSUSE Leap 15.4
 - Version [e.g. 22] :  0.17.0

**Additional context**
I want to do this for add a signature (even if this is a picture )  on a specific place of PDF.

**Comment by MatthiasValvekens** [comment]
I tried this on your file, but I can't reproduce the behaviour you report. Please run with `--verbose` and post the resulting stack trace.

**Reply by ljoli** [reply]
Hi matthias,

Thanks for your answer. I do the command with verbose option :

```
~/script_pdf/hanko> python3.10 -m pyhanko --verbose sign addfields --field  1/1,1,1,1/Sign python-cheatsheetsH.pdf  output-21.pdf
2023-02-22 07:16:40,550 - root - DEBUG - Running with --verbose
2023-02-22 07:16:40,551 - root - DEBUG - There was no configuration to parse.
2023-02-22 07:16:40,575 - pyhanko.cli - ERROR - Generic processing error.
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/pyhanko/sign/fields.py", line 1394, in prepare_sig_field
    fields = form['/Fields']
  File "/usr/lib/python3.10/site-packages/pyhanko/pdf_utils/generic.py", line 1116, in __getitem__
    raw_obj = dict.__getitem__(self, key)
KeyError: '/Fields'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/pyhanko/cli.py", line 98, in pyhanko_exception_manager
    yield
  File "/usr/lib/python3.10/site-packages/pyhanko/cli.py", line 1330, in add_sig_field
    fields.append_signature_field(writer, spec)
  File "/usr/lib/python3.10/site-packages/pyhanko/sign/fields.py", line 1563, in append_signature_field
    field_created, sig_field_ref = prepare_sig_field(
  File "/usr/lib/python3.10/site-packages/pyhanko/sign/fields.py", line 1396, in prepare_sig_field
    raise PdfError('/AcroForm has no /Fields')
pyhanko.pdf_utils.misc.PdfError: /AcroForm has no /Fields
Error: Generic processing error.
exit 1
```

After download the origin PDF from link, I have no error  : 
```
~/script_pdf> python3.10 -m pyhanko --verbose sign addfields --field  1/1,1,1,1/Sign python-cheatsheets--01.pdf  output-21.pdf 
2023-02-22 07:23:01,830 - root - DEBUG - Running with --verbose
2023-02-22 07:23:01,830 - root - DEBUG - There was no configuration to parse.
```
But I have some questions : 

1.  How to see the signature on PDF reader ? => check if PDF  has a correct signature 
2.  How to add picture like a signature on a specific position ? 
3.  is it working for all version of ref of PDF  ? => https://opensource.adobe.com/dc-acrobat-sdk-docs/pdflsdk/index.html

( I don't know if i open new discussion )




///////////////

**Reply by MatthiasValvekens** [reply]
> How to see the signature on PDF reader ? => check if PDF has a correct signature

Well, you'll have to put in an actual signature first (not just a field!), which requires having a key and a certificate to produce signatures with. See here: https://pyhanko.readthedocs.io/en/0.17.0/cli-guide/signing.html#creating-simple-signatures. 

> How to add picture like a signature on a specific position ?

PyHanko doesn't do picture-only signatures, but if you want your actual digital signature to be represented by a picture, that can be arranged. See [here](https://pyhanko.readthedocs.io/en/0.17.0/cli-guide/signing.html#customising-signature-appearances) for some basic pointers.

> is it working for all version of ref of PDF ?

Digital signatures have been a feature of PDF since version 1.3. Different PDF versions support different kinds of digital signature, though. PyHanko supports pretty much everything that's not deprecated in PDF 2.0 (+ extension standards). I generally only care about preserving PDF 1.7 and PDF 2.0 compatibility, but for most basic things, running pyHanko against an older PDF will work just fine.
